src/components/ShareListDialog.jsx

```diff
@@ -0,0 +1,81 @@
+import React, { useEffect, useRef } from 'react';
+import './ShareListDialog.css';
+
+const NO_OP = () => {};
+
+const DEFAULT_CLASSNAMES = {
+	content: 'Dialog--content',
+	form: 'Dialog--container',
+	root: 'Dialog',
+};
+
+/**
+ * @param {Omit<React.DialogHTMLAttributes<HTMLDialogElement>, 'className'> & {classNames: Partial<typeof DEFAULT_CLASSNAMES>}} props
+ */
+export function ShareListDialog({
+	children,
+	classNames,
+	open = false,
+	onCancel = NO_OP,
+	onSubmit = NO_OP,
+}) {
+	// Merge the default classNames with the user-provided ones.
+	classNames = React.useMemo(
+		() => ({ ...DEFAULT_CLASSNAMES, ...classNames }),
+		[classNames],
+	);
+
+	const ref = useRef(/** @type {?HTMLDialogElement} */ (null));
+
+	/**
+	 * When the backdrop element is clicked, close the dialog
+	 * and emit the 'cancel' event.
+	 * @param {React.MouseEvent<HTMLDialogElement, Event>} evt
+	 */
+	function handleBackdropClick(evt) {
+		if (evt.target === ref.current) {
+			ref.current.close();
+			onCancel.call(evt.target, evt);
+		}
+	}
+
+	/**
+	 * Allow the dialog to be closed by pressing the Escape key.
+	 * @param {KeyboardEvent} evt
+	 */
+	function handleKeydownCapture(evt) {
+		if (evt.key === 'Escape') {
+			ref.current.close();
+			onCancel.call(ref.current, evt);
+		}
+	}
```

Comment:
🔍 `<dialog>` elements can natively be closed with the `Escape` key. We need to extend this functionality to make sure the `'cancel'` event is called, so that any user-provided cleanup functions can run.

Reply:
This is very cool! Super interesting to see all of the functionality that takes into account keyboard only users. Thanks for this @mxmason